bpe.go

```diff
@@ -173,6 +190,101 @@ func ReadModel(reader io.Reader) (*Model, error) {
 		return &Model{}, err
 	}
 	specials, err := binaryToSpecialTokens(specialTokensBuf)
+	if err != nil {
+		return model, err
+	}
 	model.specialTokens = specials
 	return model, err
 }
+
+// IDToToken returns string token corresponding to the given token id.
+// If replaceSpace is true, special space token that is used for marking starts of words
+// will be replaced with space.
+func (m Model) IDToToken(id TokenID, replaceSpace bool) (string, error) {
+	if _, ok := m.recipe[id]; !ok {
+		if id == TokenID(m.specialTokens.unk) {
+			return unkToken, nil
+		}
+		if id == TokenID(m.specialTokens.pad) {
+			return padToken, nil
+		}
+		if id == TokenID(m.specialTokens.bos) {
+			return bosToken, nil
+		}
+		if id == TokenID(m.specialTokens.eos) {
+			return eosToken, nil
+		}
+		logrus.Errorf("%d: token id is out of vocabulary", id)
+		return "", errors.New("token id is out of vocabulary")
+	}
+	encodedToken, _ := m.recipe[id]
+	if encodedToken[0] == m.spaceID && replaceSpace {
+		token, err := DecodeToken(encodedToken[1:], m.id2char)
+		if err != nil {
+			return "", err
+		}
+		return " " + token, nil
+	}
+	return DecodeToken(encodedToken, m.id2char)
+}
+
+// DecodeSentence decodes a sequence of token ids in a text sentence - string of words
+// with spaces in between
+func (m Model) DecodeSentence(encodedSentence EncodedString) (string, error) {
+	sentence := ""
+	for _, tokenID := range encodedSentence {
+		token, err := m.IDToToken(tokenID, true)
+		if err != nil {
+			return sentence, err
+		}
+		sentence += token
+	}
+	if string(sentence[0]) == " " {
+		sentence = sentence[1:]
+	}
+	if sentence[:len(bosToken)+1] == bosToken+" " {
+		sentence = bosToken + sentence[len(bosToken)+1:]
+	}
+	return sentence, nil
+}
+
+// DecodeSentences decodes a sequence of encoded sentences - sequences of token ids -
+// into a sequence of corresponding text sentences
+func (m Model) DecodeSentences(encodedSentences []EncodedString) ([]string, error) {
+	sentences := make([]string, len(encodedSentences))
+	for i, encodedSentence := range encodedSentences {
+		sentence, err := m.DecodeSentence(encodedSentence)
+		if err != nil {
+			return sentences, err
+		}
+		sentences[i] = sentence
+	}
+	return sentences, nil
+}
+
+// DecodeFromStream decodes a sequence of encoded sentences written in an input stream
+// using Model.DecodeSentences
+func (m Model) DecodeFromStream(reader io.Reader) ([]string, error) {
+	scanner := bufio.NewScanner(reader)
+	sentences := make([]string, 0)
```

Comment:
```suggestion
	var sentences []string
```

`append` works with `nil` slices and this is considered idiomatic. Besides, `len()` also works for nil slices.

However, inserting to `nil` maps does not work!